core/src/main/java/kafka/server/DefaultAutoTopicCreationManager.java

```diff
@@ -0,0 +1,274 @@
+                        .setErrorCode(validationError.get().code())
+                        .setName(topic)
+                        .setIsInternal(Topic.isInternal(topic)));
+            } else {
+                creatableTopics.put(topic, creatableTopic(topic));
+            }
+        });
+        var creatableTopicResponses = creatableTopics.isEmpty() ?
+                List.<MetadataResponseTopic>of() : sendCreateTopicRequest(creatableTopics, metadataRequestContext);
+        return Stream.concat(uncreatableTopicResponses.stream(), creatableTopicResponses.stream())
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public void createStreamsInternalTopics(
+            Map<String, CreatableTopic> topics,
+            RequestContext requestContext
+    ) {
+        if (topics.isEmpty()) {
+            return;
+        }
+        topics.values().forEach(creatableTopic -> {
+            if (creatableTopic.numPartitions() == -1) {
+                creatableTopic.setNumPartitions(config.numPartitions());
+            }
+            if (creatableTopic.replicationFactor() == -1) {
+                creatableTopic.setReplicationFactor((short) config.defaultReplicationFactor());
+            }
+        });
+        sendCreateTopicRequest(topics, Optional.of(requestContext));
+    }
+
+    private List<MetadataResponseTopic> sendCreateTopicRequest(
+            Map<String, CreatableTopic> creatableTopics,
+            Optional<RequestContext> requestContext
+    ) {
+        var topicsToCreate = new CreateTopicsRequestData.CreatableTopicCollection(creatableTopics.values().iterator());
+        var createTopicsRequest = new CreateTopicsRequest.Builder(
+                new CreateTopicsRequestData()
+                        .setTimeoutMs(config.requestTimeoutMs())
+                        .setTopics(topicsToCreate)
+        );
+
+        var requestCompletionHandler = new ControllerRequestCompletionHandler() {
+            @Override
+            public void onTimeout() {
+                clearInflightRequests(creatableTopics);
+                LOGGER.debug("Auto topic creation timed out for {}.", creatableTopics.keySet());
+            }
+
+            @Override
+            public void onComplete(ClientResponse response) {
+                clearInflightRequests(creatableTopics);
+                if (response.authenticationException() != null) {
+                    LOGGER.warn("Auto topic creation failed for {} with authentication exception.", creatableTopics.keySet());
+                } else if (response.versionMismatch() != null) {
+                    LOGGER.warn("Auto topic creation failed for {} with invalid version exception.", creatableTopics.keySet());
+                } else {
+                    if (response.hasResponse()) {
+                        if (response.responseBody() instanceof CreateTopicsResponse) {
+                            var createTopicsResponse = (CreateTopicsResponse) response.responseBody();
```

Comment:
We could leverage pattern matching here.
```suggestion
                        if (response.responseBody() instanceof CreateTopicsResponse createTopicsResponse) {
```